channels/rdpgfx/client/rdpgfx_main.c

```diff
@@ -668,14 +598,299 @@ static UINT rdpgfx_recv_evict_cache_entry_pdu(RDPGFX_CHANNEL_CALLBACK* callback,
+	             pdu->cacheEntriesCount);
+	s = Stream_New(NULL, header.pduLength);
+
+	if (!s)
+	{
+		WLog_ERR(TAG, "Stream_New failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	if ((error = rdpgfx_write_header(s, &header)))
+		goto fail;
+
+	if (pdu->cacheEntriesCount <= 0)
+	{
+		WLog_ERR(TAG, "Invalid cacheEntriesCount: %" PRIu16 "", pdu->cacheEntriesCount);
+		error = ERROR_INVALID_DATA;
+		goto fail;
+	}
+
+	/* cacheEntriesCount (2 bytes) */
+	Stream_Write_UINT16(s, pdu->cacheEntriesCount);
+
+	for (index = 0; index < pdu->cacheEntriesCount; index++)
+	{
+		cacheEntry = &(pdu->cacheEntries[index]);
+		Stream_Write_UINT64(s, cacheEntry->cacheKey);     /* cacheKey (8 bytes) */
+		Stream_Write_UINT32(s, cacheEntry->bitmapLength); /* bitmapLength (4 bytes) */
+	}
+
+	error = callback->channel->Write(callback->channel, (UINT32)Stream_Length(s), Stream_Buffer(s),
+	                                 NULL);
+
+fail:
+	Stream_Free(s, TRUE);
+	return error;
+}
+
+static int rdpgfx_send_cache_offer(RDPGFX_PLUGIN* gfx)
+{
+	int idx, count;
+	PERSISTENT_CACHE_ENTRY entry;
+	RDPGFX_CACHE_IMPORT_OFFER_PDU offer;
+	rdpPersistentCache* persistent = NULL;
+	RdpgfxClientContext* context = (RdpgfxClientContext*)gfx->iface.pInterface;
+	rdpSettings* settings = gfx->settings;
+
+	if (!settings->BitmapCachePersistEnabled)
+		return 0;
+
+	if (!settings->BitmapCachePersistFile)
+		return 0;
+
+	ZeroMemory(&offer, sizeof(RDPGFX_CACHE_IMPORT_OFFER_PDU));
+
+	persistent = persistent_cache_new();
+
+	if (!persistent)
+		return -1;
+
+	if (persistent_cache_open(persistent, settings->BitmapCachePersistFile, FALSE, 3) < 1)
+		goto error;
+
+	if (persistent_cache_get_version(persistent) != 3)
+		goto error;
+
+	count = persistent_cache_get_count(persistent);
+
+	if (count >= RDPGFX_CACHE_ENTRY_MAX_COUNT)
+		count = RDPGFX_CACHE_ENTRY_MAX_COUNT - 1;
+
+	if (count > gfx->MaxCacheSlots)
+		count = gfx->MaxCacheSlots;
+
+	offer.cacheEntriesCount = (UINT16) count;
+
+	WLog_DBG(TAG, "Sending Cache Import Offer: %d", count);
+
+	for (idx = 0; idx < count; idx++)
+	{
+		persistent_cache_read_entry(persistent, &entry);
+		offer.cacheEntries[idx].cacheKey = entry.key64;
+		offer.cacheEntries[idx].bitmapLength = entry.size;
+	}
+
+	persistent_cache_free(persistent);
+
+	if (offer.cacheEntriesCount > 0) {
+		if (rdpgfx_send_cache_import_offer_pdu(context, &offer) != CHANNEL_RC_OK)
+		{
+			WLog_Print(gfx->log, WLOG_ERROR, "Failed to send cache import offer PDU");
+			goto error;
+		}
+	}
+
+	return 1;
+error:
+	persistent_cache_free(persistent);
+	return -1;
+}
+
+static int rdpgfx_load_cache_import_reply(RDPGFX_PLUGIN* gfx, RDPGFX_CACHE_IMPORT_REPLY_PDU* reply)
+{
+	int idx;
+	int count;
+	UINT16 cacheSlot;
+	PERSISTENT_CACHE_ENTRY entry;
+	rdpPersistentCache* persistent = NULL;
+	rdpSettings* settings = gfx->settings;
+	RdpgfxClientContext* context = (RdpgfxClientContext*)gfx->iface.pInterface;
+
+	if (!settings->BitmapCachePersistEnabled)
```

Comment:
Repeated checks (`RDPGFX_PLUGIN* gfx` valid, `settings` proper, ...)
maybe unify in a static function doing all these checks?